pool/client_test.go

```diff
@@ -542,78 +938,218 @@ func testClient(t *testing.T, db *bolt.DB) {
+	if !ok {
+		t.Fatalf("unable to cast message as request")
+	}
+	if req.Method != SetDifficulty {
+		t.Fatalf("expected %s message method, got %s", SetDifficulty, req.Method)
+	}
+
+	// Ensure a CPU client receives a valid non-error response when
+	// a valid subscribe request is sent.
+	id++
+	r = SubscribeRequest(&id, "mcpu", "1.0.1", "")
+	err = sE.Encode(r)
+	if err != nil {
+		t.Fatalf("[Encode] unexpected error: %v", err)
+	}
+	data = <-recvCh
+	msg, mType, err = IdentifyMessage(data)
+	if err != nil {
+		t.Fatalf("[IdentifyMessage] unexpected error: %v", err)
+	}
+	if mType != ResponseMessage {
+		t.Fatalf("expected a subscribe response message, got %v", mType)
+	}
+	resp, ok = msg.(*Response)
+	if !ok {
+		t.Fatalf("expected subsribe response with id %d, got %d", *r.ID, resp.ID)
+	}
+	if resp.ID != *r.ID {
+		t.Fatalf("expected suscribe response with id %d, got %d", *r.ID, resp.ID)
+	}
+	if resp.Error != nil {
+		t.Fatalf("expected non-error subscribe response, got %v", resp.Error)
+	}
+
+	// Ensure the CPU client is now authorized and subscribed
+	// for work updates.
+	client.authorizedMtx.Lock()
+	authorized = client.authorized
+	client.authorizedMtx.Unlock()
+
+	if !authorized {
+		t.Fatalf("expected an authorized mining client")
+	}
+
+	client.subscribedMtx.Lock()
+	subscribed = client.subscribed
+	client.subscribedMtx.Unlock()
+
+	if !subscribed {
+		t.Fatalf("expected a subscribed mining client")
+	}
+
+	// Trigger time-rolled work updates to the CPU  client.
+	setCurrentWork(workE)
+
+	// Send a work notification to the CPU client.
+	r = WorkNotification(job.UUID, prevBlock, genTx1, genTx2,
+		blockVersion, nBits, nTime, true)
+	client.ch <- r
+	cpuWork = <-recvCh
+	msg, mType, err = IdentifyMessage(cpuWork)
+	if err != nil {
+		t.Fatalf("[IdentifyMessage] unexpected error: %v", err)
+	}
+	if mType != NotificationMessage {
+		t.Fatalf("expected a notification message, got %v", mType)
+	}
+	req, ok = msg.(*Request)
+	if !ok {
+		t.Fatalf("unable to cast message as request")
+	}
+	if req.Method != Notify {
+		t.Fatalf("expected %s message method, got %s", Notify, req.Method)
+	}
+
 	// Ensure the client receives time-rolled work.
 	timeRolledWork := <-recvCh
 	msg, mType, err = IdentifyMessage(timeRolledWork)
 	if err != nil {
 		t.Fatalf("[IdentifyMessage] unexpected error: %v", err)
 	}
-
 	if mType != NotificationMessage {
 		t.Fatalf("expected a notification message, got %v", mType)
 	}
-
 	req, ok = msg.(*Request)
 	if !ok {
 		t.Fatalf("unable to cast message as request")
 	}
-
 	if req.Method != Notify {
 		t.Fatalf("expected %s message method, got %s", Notify, req.Method)
 	}
 
-	// Fake a bunch of submissions and calculate the hash rate.
-	setMiner(CPU)
-	atomic.StoreInt64(&client.submissions, 50)
-	time.Sleep(time.Second * 2)
-	hash := client.fetchHashRate()
-	if hash == ZeroRat {
-		t.Fatal("expected a non-nil client hash rate")
-	}
+	// Trigger a client timeout by waiting.
+	time.Sleep(time.Second * 3)
```

Comment:
Why is this required?

Reply:
Idling here triggers the connection timeout.